## Machine Learning - CA2.2
## Matthew Riddell - D00245674

For this CA, I wish to look at do a simple classification between Cats and Dogs 
I selected the following Dataset from kaggle:

https://www.kaggle.com/datasets/tongpython/cat-and-dog



### This code will also be uploaded to my github repo here:

https://github.com/Matthew-Riddell/Machine-Learning-CA2.2


Code is adapted from module exercises and from tensorflow documentation

#### Package Imports

In [20]:
import tensorflow as tf
from tensorflow.keras import layers, models
import matplotlib.pyplot as plt
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense
from tensorflow.keras.optimizers import Adam

#### Dataset folders

In [6]:
datasets_folder = "C:/Users/Matty/Documents/College Notes & Assignments/Year 5/Machine Learning/CAs/CA2.2/Machine-Learning-CA2.2/dataset/"
train_folder = datasets_folder + "training_set/training_set/"
test_folder = datasets_folder + "test_set/test_set/"

#### Loading the dataset and dividing the images into batches

loads the images from the folders and keras applies labels

resizes the images to 128x128 and takes them in batches of 32 images at a time

image_dataset_from_directory() then returns the count of x (images) and y (classes) which are cats and dogs 

In [ ]:
# https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/image_dataset_from_directory

train_data = tf.keras.preprocessing.image_dataset_from_directory(
    train_folder,
    image_size=(128, 128),
    batch_size=32
)

test_data = tf.keras.preprocessing.image_dataset_from_directory(
    test_folder,
    image_size=(128, 128),
    batch_size=32
)

Found 8005 files belonging to 2 classes.
Found 2023 files belonging to 2 classes.


#### Normalization

rescaling the pixels to be values between 0 and 1 so they work better in the neural network

the normalization is then applied to each image in the datasets via the x (image) and y (class) values 

In [10]:
# https://www.tensorflow.org/api_docs/python/tf/keras/layers/Rescaling

normalization = layers.Rescaling(1./255)
train_data = train_data.map(lambda x, y: (normalization(x), y))
test_data = test_data.map(lambda x, y: (normalization(x), y))

### Simple CNN Model

#### Making the CNN Model

here i make a simple sequential CNN model like so:

Input layer -> Convolutional Layer 1 -> Pooling Layer 1 -> Convolutional Layer 2 -> Pooling Layer 2 -> Flatten Layer -> Dense Layer -> Output Layer



In [14]:
# https://www.tensorflow.org/guide/keras/sequential_model?hl=en

model = Sequential()

# Convolutional Layer 1
model.add(Conv2D(32, (3,3), activation='relu', input_shape=(128,128,3)))

# Pooling Layer 1
model.add(MaxPooling2D(pool_size=(2,2)))

# Convolutional Layer 2
model.add(Conv2D(64, (3,3), activation='relu'))

# Pooling Layer 2
model.add(MaxPooling2D(pool_size=(2,2)))

# Flatten Layer
model.add(Flatten())

# Dense Layer
model.add(Dense(64, activation='relu'))

# Output Layer
model.add(Dense(1, activation='sigmoid'))

#### Model shape and size

In [17]:
model.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_4 (Conv2D)               │ (None, 126, 126, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_4 (MaxPooling2D)  │ (None, 63, 63, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_5 (Conv2D)               │ (None, 61, 61, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_5 (MaxPooling2D)  │ (None, 30, 30, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_2 (Flatten)             │ (None, 57600)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 64)             │     3,686,464 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,705,921 (14.14 MB)

 Trainable params: 3,705,921 (14.14 MB)

 Non-trainable params: 0 (0.00 B)

#### Compiling the Model

preparing the model for training using the adam optimizer and using binary cross entropy as the loss function since it is a binary classification

In [ ]:
# https://www.tensorflow.org/api_docs/python/tf/keras/Model
# https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam

model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

#### Training the Model

training this first model on 5 epoch

In [18]:
model.fit(train_data, epochs=5)

Epoch 1/5
251/251 ━━━━━━━━━━━━━━━━━━━━ 71s 273ms/step - accuracy: 0.4979 - loss: 0.6936
Epoch 2/5
251/251 ━━━━━━━━━━━━━━━━━━━━ 79s 312ms/step - accuracy: 0.4914 - loss: 0.6932
Epoch 3/5
251/251 ━━━━━━━━━━━━━━━━━━━━ 62s 247ms/step - accuracy: 0.4987 - loss: 0.6932
Epoch 4/5
251/251 ━━━━━━━━━━━━━━━━━━━━ 92s 287ms/step - accuracy: 0.4947 - loss: 0.6932
Epoch 5/5
251/251 ━━━━━━━━━━━━━━━━━━━━ 72s 285ms/step - accuracy: 0.4937 - loss: 0.6932


In [19]:
loss, acc = model.evaluate(test_data)
print("Test accuracy:", acc)

64/64 ━━━━━━━━━━━━━━━━━━━━ 6s 82ms/step - accuracy: 0.4998 - loss: 0.6932
Test accuracy: 0.49975284934043884


accuracy: 0.4998 - loss: 0.6932

very mediocre performance for this simple CNN

it is basically correct 50% of the time

### Tuned Simple CNN Model